src/hepattn/utils/sorting.py

```diff
@@ -0,0 +1,127 @@
+        assert key_sort_idx.shape[0] == self.num_hits, f"Key sort index shape {key_sort_idx.shape} does not match num_hits {self.num_hits}"
+        self.sort_indices["key"] = key_sort_idx
+
+        # TODO: sort key_phi (key_{input_sort_field})
+        x[f"key_{self.input_sort_field}"] = self._sort_tensor_by_index(x[f"key_{self.input_sort_field}"], key_sort_idx, self.num_hits)
+
+        for input_name in [input_net.input_name for input_net in self.input_nets]:
+            if input_name + "_embed" in x:
+                x[input_name + "_embed"] = self._sort_tensor_by_index(x[input_name + "_embed"], key_sort_idx, self.num_hits)
+            if input_name + "_valid" in x:
+                x[input_name + "_valid"] = self._sort_tensor_by_index(x[input_name + "_valid"], key_sort_idx, self.num_hits)
+            if f"key_is_{input_name}" in x:
+                x[f"key_is_{input_name}"] = self._sort_tensor_by_index(x[f"key_is_{input_name}"], key_sort_idx, self.num_hits)
+
+        # Sort key embeddings
+        x["key_embed"] = self._sort_tensor_by_index(x["key_embed"], key_sort_idx, self.num_hits)
+
+        # Sort key validity mask
+        if x["key_valid"] is not None:
+            x["key_valid"] = self._sort_tensor_by_index(x["key_valid"], key_sort_idx, self.num_hits)
+
+        # Sort raw variables if they have the right shape
+        for raw_var in self.raw_variables:
+            if raw_var in x and x[raw_var].shape[-1] == x["key_embed"].shape[-2]:
+                x[raw_var] = self._sort_tensor_by_index(x[raw_var], key_sort_idx, self.num_hits)
+            else:
+                print(f"Warning: Raw variable {raw_var} has invalid shape: {x[raw_var].shape}")
+
+        return x
+
+    def sort_targets(self, targets: dict, sort_indices: dict[str, Tensor]) -> dict:
+        """Sort targets to align with sorted outputs."""
+        # TODO: check that this sorts all the targets correctly
+        for key, value in targets.items():
+            targets[key] = self._sort_tensor_by_index(value, sort_indices["key"], self.num_hits)
+
+        return targets
+
+    def _sort_tensor_by_index(self, tensor: Tensor, sort_idx: Tensor, num_hits: int) -> Tensor:
+        """Sort a tensor along the dimension that has the same shape as key_embed[0].
+
+        Parameters
+        ----------
+        tensor : Tensor
+            Tensor to sort.
+        sort_idx : Tensor
+            Sort indices.
+        num_hits : int
+            Number of hits.
+
+        Returns:
+        Tensor
+            Sorted tensor.
+        """
+        if tensor is None:
+            return None
+        sort_dim = None
+        # If num hits is provided, find the dimension with matching size
+        for dim, size in enumerate(tensor.shape):
+            if size == num_hits:
+                sort_dim = dim
+                break
+
+        if sort_dim is not None:
+            return tensor.index_select(sort_dim, sort_idx.to(tensor.device))
```

Comment:
Let's assume the tensor is already on the correct device and remove `.to(tensor.device)`